pkg/image/prune/imagepruner.go

```diff
@@ -25,42 +27,147 @@ import (
-// prunable.
-type ImagePruneFunc func(image *imageapi.Image) error
-type ImageStreamPruneFunc func(stream *imageapi.ImageStream, image *imageapi.Image) (*imageapi.ImageStream, error)
-type LayerPruneFunc func(registryURL, repo, layer string) error
-type BlobPruneFunc func(registryURL, blob string) error
-type ManifestPruneFunc func(registryURL, repo, manifest string) error
-
-// ImagePruner knows how to prune images and layers.
+// ImagePruner knows how to delete images from OpenShift.
 type ImagePruner interface {
-	// Run prunes images and layers.
-	Run(pruneImage ImagePruneFunc, pruneStream ImageStreamPruneFunc, pruneLayer LayerPruneFunc, pruneBlob BlobPruneFunc, pruneManifest ManifestPruneFunc)
+	// PruneImage deletes the image from OpenShift's storage.
+	PruneImage(image *imageapi.Image) error
+}
+
+// ImageStreamPruner knows how to remove an image reference from an image
+// stream.
+type ImageStreamPruner interface {
+	// PruneImageStream deletes all references to the image from the image
+	// stream's status.tags. The updated image stream is returned.
+	PruneImageStream(stream *imageapi.ImageStream, image *imageapi.Image, updatedTags []string) (*imageapi.ImageStream, error)
+}
+
+// BlobPruner knows how to delete a blob from the Docker registry.
+type BlobPruner interface {
+	// PruneBlob uses registryClient to ask the registry at registryURL to delete
+	// the blob.
+	PruneBlob(registryClient *http.Client, registryURL, blob string) error
+}
+
+// LayerPruner knows how to delete a repository layer link from the Docker
+// registry.
+type LayerPruner interface {
+	// PruneLayer uses registryClient to ask the registry at registryURL to
+	// delete the repository layer link.
+	PruneLayer(registryClient *http.Client, registryURL, repo, layer string) error
+}
+
+// ManifestPruner knows how to delete image manifest data for a repository from
+// the Docker registry.
+type ManifestPruner interface {
+	// PruneManifest uses registryClient to ask the registry at registryURL to
+	// delete the repository's image manifest data.
+	PruneManifest(registryClient *http.Client, registryURL, repo, manifest string) error
+}
+
+// ImageRegistryPrunerOptions contains the fields used to initialize a new
+// ImageRegistryPruner.
+type ImageRegistryPrunerOptions struct {
+	KeepYoungerThan  time.Duration
+	KeepTagRevisions int
+	Images           *imageapi.ImageList
+	Streams          *imageapi.ImageStreamList
+	Pods             *kapi.PodList
+	RCs              *kapi.ReplicationControllerList
+	BCs              *buildapi.BuildConfigList
+	Builds           *buildapi.BuildList
+	DCs              *deployapi.DeploymentConfigList
+	DryRun           bool
+	RegistryClient   *http.Client
+	RegistryURL      string
 }
 
-// imagePruner implements ImagePruner.
-type imagePruner struct {
-	g         graph.Graph
-	algorithm pruneAlgorithm
+// ImageRegistryPruner knows how to prune images and layers.
+type ImageRegistryPruner interface {
+	// Prune prunes images and layers.
```

Comment:
@pmorie updated. Let me know if this is any better. I considered moving the comment from NewImageRegistryPruner here, but the actual algorithm is part of imageRegistryPruner (the struct), not ImageRegistryPruner (the interface).